### DISCUSSION TITLE
How to create a plugin for `.parse` only

### DISCUSSION THREAD

**Discussion by jeffsee55** [remarkjs/remark#1014] [Q&A] [answered]
I'm working with the AST of a markdown file so I'm using `.parse`, I've been searching for examples but every thing I see is an example of `.process` (eg. https://github.com/remarkjs/remark-breaks). From what I understand, `.process` is the whole thing `string -> AST -> string`, while `.parse` is just `string -> AST`. I'd like to extend the parsing logic from remark so that I can tweak some of the text nodes and pass it on to the rest of my business logic.

Every example I can find uses `.process`, and the API looks simple but doesn't run with `.parse`

In a nutshell, I'd like to understand why this logs the tree:

```js
import { remark } from "remark";
remark()
  .use(() => (tree) => {
    console.log(tree);
  })
  .process("some text");

// logs { type: 'root', children: [{ ... 
```

And this logs nothing:

```js
import { remark } from "remark";
remark()
  .use(() => (tree) => {
    console.log(tree);
  })
  .parse("some text");
```

**Comment by wooorm** [comment, marked as answer]
That is not what plugins are for.
For the background, read the readme of unified until the API section: https://github.com/unifiedjs/unified.

**Reply by ChristianMurphy** [reply]
To add on to what @wooorm is saying.
Transformer (AST based) plugins only run during `process` and `run` by design.

```
| ........................ process ........................... |
| .......... parse ... | ... run ... | ... stringify ..........|

          +--------+                     +----------+
Input ->- | Parser | ->- Syntax Tree ->- | Compiler | ->- Output
          +--------+          |          +----------+
                              X
                              |
                       +--------------+
                       | Transformers |
                       +--------------+
```

https://github.com/unifiedjs/unified#overview

That said, you can manually run an AST through a plugin if you want with:

```js
import { remark } from "remark"

const remarkPlugin = () => (tree) => { console.log(tree) }

// Option 1: with parse and run
const pipeline = remark().use(remarkPlugin)
const ast = pipeline
  .parse("some text")
pipeline.run(ast)

// Option 2: with parse and manually call
const ast = remark()
  .parse("some text")

remarkPlugin({})(ast)
```

and the same goes for other transformer plugins, like `remark-breaks`

**Reply by jeffsee55** [reply]
Thanks @wooorm, I have given that a read but it's a little overwhelming. Could you expand on what you mean by "not what plugins are for"? Manipulating the AST seems like a normal use-case from what I understood. So for a plugin like this https://github.com/mrzmmr/remark-behead, it alters the AST but there's no way to return it I guess? 

It sounds like I need to work exclusively with the lower level `mdast-` level. Is there a better place for me to go with this question?

**Reply by wooorm** [reply]
See the diagram and embrace unified.

You should likely use remark/unified.

But you should let it do what it’s for: taking a string and parsing it, changing and inspecting that tree, and then serializing it. Instead of trying to do each step manually.

**Reply by jeffsee55** [reply]
We're actually serving the AST over a network call as part of a CMS. So it's similar to other headless CMSs in that way, but with the benefits of working with markdown, so we don't want to serialize anything. (the project is https://github.com/tinacms/tinacms).

**Reply by wooorm** [reply]
If you don’t want 3 steps, you can use the 2 steps you want: call `.parse` and `.run`.

Serializing the AST can also be seen as serializing. You can add a small compiler:

```js
function myUnifiedPluginThatSerializezTheAst() {
  this.Compiler = (tree) => JSON.stringify(tree)
}
```

**Reply by ChristianMurphy** [reply]
An AST really isn't a wire protocol (see https://github.com/unifiedjs/unified/discussions/188#discussioncomment-2466114), trying to use it as such may not offer many (if any) benefits.
You're welcome to try anyway, see my reply from earlier with examples of how you can run the pipeline without the serialization step, or @wooorm's `JSON.stringify` serializer above.

**Reply by jeffsee55** [reply]
Will give the `.run` a shot, thanks for the tip. 

For what it's worth the AST is simple enough to make it comparable to other headless CMS rich-text offerings. We haven't had any issue with regard to payload size but perhaps I'm missing something?

**Reply by ChristianMurphy** [reply]
There are two main uses cases I've seen in CMSes that leverage `remark` (or markdown in general).
1. viewing a static piece of content
2. editing a piece of content

For the first, sending static content either: sending the raw markdown (and letting the frontend do everything to take pressure off the backend) or sending fully processed HTML (letting the backend do all the work, and the frontend reap the benefits) makes sense.
Sending the AST seems like an unhappy medium (still loading a bunch of JavaScript on the front end, also with an AST payload which is about the same size as the final HTML one would be).

In the second, editing content, you need to be able to parse Markdown on the frontend anyway, so why add complexity to the backend? Let the frontend parse and preview.

That's not to say there _coundn't_ be a reason to use an AST as a wire protocol.
But in the vast majority of cases, it is overkill, adds complexity, and may hurt performance.

**Reply by jeffsee55** [reply]
Yeah I see what you mean, but as a CMS we don't really mix with the user's code, so it's our job to serve the JSON and let the user decide what to do with it. This is what makes the AST a solid compromise. They don't need to bring their own serialization step, and they'll most often still render it via SSR anyway (and the payload from us easy to cache). We're very much in the Next.js space right now so it's a good fit.

> In the second, editing content, you need to be able to parse Markdown on the frontend anyway

For editing the content we provide a rich-text editor, which uses the AST so we don't ship any of that logic to the frontend. We _also_ provide a "renderer" for React, admittedly without that component this would all be pretty tough to do. (Note that we have tweaked the AST to work better with Slate)

Rough example here https://tina-gql-playground.vercel.app/rich-text
